Initial development (CORE-9175)

ianmcorvidae opened the pull request:
This is the initial development of a new infosquito. It works by a different means than the previous iteration -- instead of first doing a pass over the whole elasticsearch index, checking as it goes that everything exists and deleting from the index what doesn't, then a second pass over the entire ICAT database, adding to elasticsearch unconditionally, this code:

* divides the ID space (uuids) into partitions by a configurable number of initial digits (with our production database, four digits works well, but I've set the default to 3 because that will probably be better with smaller databases)
* if running in the AMQP mode (`-mode periodic`) it sends messages for each prefix to be processed by itself or other instances; if in `-mode full` it simply loops over them
* for each prefix, it requests everything from elasticsearch and checks how many results are returned (and similarly for the DB); if the number is too large (by default 10000, based on elasticsearch's default for the maximum returned results from a query) it splits the prefix further and either iterates or sends messages as before
* it then requests every data object from the DB; as it goes through the data objects, it adds anything whose ID does not appear in elasticsearch or whose JSON differs meaningfully (this comparison is in `document.go`), then repeats this for collections, keeping track of which IDs it has seen
* Once done, it deletes anything from the elasticsearch index which was not seen in either of the database queries

On testing, with my laptop close to the DB and elasticsearch, it took about 22 hours for `-mode full` and with it relatively far away (at home, over VPN and over an SSH tunnel for the DB), it took 12 hours and 45 minutes for `-mode periodic` with four instances. These numbers can of course vary more, since this code will do less work if there's less that needs to be done, but it seems like a pretty respectable baseline.

ianmcorvidae commented on main.go:
```diff
@@ -0,0 +1,252 @@
+	return res
+}
+
+func splitPrefix(prefix string) []string {
```
This would stop working if we got up to 8, but we can deal with that when it becomes relevant.

tedgin replied:
Sorry for being dense here, but 8 of what would break this?

ianmcorvidae replied:
This function will generate prefixes that match nothing if it's passed an 8-character prefix, because the ninth character of a UUID is `-`, not a hexadecimal digit.

ianmcorvidae replied:
The same applies to `generatePrefixes` as well. But it'll be quite a while (possibly effectively forever) before we should actually need prefixes anywhere near that long, so I figured I'd leave it for now.

tedgin commented on main.go:
```diff
@@ -0,0 +1,252 @@
+		amqpExchangeType,
+		queueName,
+		[]string{"index.all", "index.data", "index.data.prefix.#"},
+		func(del amqp.Delivery) {
```
This anonymous function is pretty big.  Could it be made into a named function inside the main package?

ianmcorvidae replied:
Yeah, that's probably a good idea, or maybe I'll keep it anonymous but break out bits of it. I figure since it essentially does two things (generate prefixes and send messages vs. try to reindex and send messages if needed) maybe those should be separate too.

tedgin replied:
That makes sense.

tedgin commented on main.go:
```diff
@@ -0,0 +1,252 @@
+					}
+				}
+			} else {
+				prefix := del.RoutingKey[18:]
```
This is a little brittle assuming the base routing key is going to be 18 characters long.  Maybe declare the base key as a constant an compute its length.

ianmcorvidae replied:
Yeah, that's a good idea, I'll do that!

tedgin commented on reindex.go:
```diff
@@ -0,0 +1,266 @@
+	prefixlog.Infof("Processed %d entries (%d rows, %d documents, processed %d data objects (+%d,U%d,-%d), %d colls (+%d,U%d,-%d)) in %s", rows.processed, rows.rows, rows.documents, rows.dataobjects, rows.dataobjects_added, rows.dataobjects_updated, rows.dataobjects_removed, rows.colls, rows.colls_added, rows.colls_updated, rows.colls_removed, time.Since(start).String())
+}
+
+func ReindexPrefix(db *ICATConnection, es *ESConnection, prefix string) error {
```
This should be broken up into multiple functions to make it easier to visualize the algorithm.

ianmcorvidae replied:
Yeah, another good idea. My own fault, since I effectively grew this from an empty function to... well, this, without sitting down to think about the structure. I'll give it a look.

ianmcorvidae commented:
I think I've addressed your concerns, so I'm going to merge this. If there's more you'd like me to do, I'm still on board, I just want to start on getting this deployed!
